docs/Changelog-ml.md

```diff
@@ -1147,3 +1147,84 @@ This version of the operator has been available since version 3 of the 'ai.onnx.
 <dd>The input type must be a tensor of a numeric type.</dd>
 </dl>
 
+## Version 4 of the 'ai.onnx.ml' operator set
+### <a name="ai.onnx.ml.LabelEncoder-4"></a>**ai.onnx.ml.LabelEncoder-4**</a>
+
+  Maps each element in the input tensor to another value.<br>
+      The mapping is determined by the two parallel attributes, 'keys_*' and
+      'values_*' attribute. The i-th value in the specified 'keys_*' attribute
+      would be mapped to the i-th value in the specified 'values_*' attribute. It
+      implies that input's element type and the element type of the specified
+      'keys_*' should be identical while the output type is identical to the
+      specified 'values_*' attribute. If an input element can not be found in the
+      specified 'keys_*' attribute, the 'default_*' that matches the specified
+      'values_*' attribute may be used as its output value.<br>
+      Let's consider an example which maps a string tensor to an integer tensor.
+      Assume and 'keys_strings' is ["Amy", "Sally"], 'values_int64s' is [5, 6],
+      and 'default_int64' is '-1'.  The input ["Dori", "Amy", "Amy", "Sally",
+      "Sally"] would be mapped to [-1, 5, 5, 6, 6].<br>
+      Since this operator is an one-to-one mapping, its input and output shapes
+      are the same. Notice that only one of 'keys_*'/'values_*' can be set.<br>
+      When 'keys_as_tensor' and 'values_as_tensor' are used, the default value
+      must be specified in 'default_as_tensor' attribute as a singleton with the
+      same type as 'values_as_tensor'. Additionally, 'keys_as_tensor' and
+      'values_as_tensor' must be one dimensional and have the same length.<br>
+      For key look-up, bit-wise comparison is used so even a float NaN can be
+      mapped to a value in 'values_*' attribute.<br>
```

Comment:
I think what you are saying is that the standard does not explicitly specify that `nan` values in `AttributeProto` should be serialised with their exact bit pattern (at least in the LabelEncoder). I'm not sure if protobuf provides any guarantees for with float serialisation but I think it's a valid issue with the spec as it stands.

Reply:
There are also no guarantees about which kind of `NaN` values are produced by other operators such as `Div` or `Cast`.

The previous wording was "For key look-up, bit-wise comparison is used so even a float NaN can be mapped to a value in 'values_*' attribute.". However, since the standard makes no guarantees about the exact bit patterns there has never been a reliable way to map some `NaN` bit patterns to one value and other patterns to other values. Thus, slightly changing the wording now (to arguably be in better agreement with the previous intent) is not a breaking change (that might require a deprecation cycle) because there has never been a reliable way to control the bit pattern of the inputs in the first place.

Did this make my thoughts clearer?

Reply:
I haven't thought much about this, but what does "the standard makes no guarantees about the exact bit pattern" mean? Doesn't the IEEE standard for float/double etc. prescribe this? I think so does float16, and the newly introduced 8-bit floats, etc.

Reply:
The IEEE standard allows for millions of NaN bit patterns and as far as I know it is processor dependent which pattern is used in what situation. See for example https://en.wikipedia.org/wiki/NaN#Encoding (especially the last paragraph), but please correct me if I'm wrong!

Float16 types may also have different bit patterns. Here are two example-NaNs: https://float.exposed/b0x7fc0 and https://float.exposed/b0xffff

Either way, it seems very fiddly and unpredictable to use bit patterns in the context of the `LabelEncoder`. The intent has probably all along been to map any `NaN` input to a single value, hasn't it?

Reply:
However, if normalizing NaN values is a common use-case/scenario, but not everyone wants it, one could support it via an attribute. I must confess I do not know what the various common use-cases are: if it is only one, we can fit the operator to that, but if different users want different behavior, we may need to account for that.

Reply:
Sorry for maybe missing the essence of your question in my earlier reply! The point is that the current behavior around `NaN`s for the `LabelEncoder` is unusable. There is no guarantee in the ONNX standard about preserving the bit-patterns of `NaN`s in any operation or even directly in the inputs. Matching against a particular pattern, as is currently suggested by the documentation of the `LabelEncoder`, leads to an undefined hit-and-miss behavior. It might work but is really neither reliable nor convenient to work with. I don't see the point in leaving it as is, to be honest.

So the question becomes if it should be removed, or if the specification should be updated to something workable. If it were to be removed one would indeed have to work around with `IsNaN` and `Where`. Not the nicest, but workable

The alternative is the suggested "upgrade" where all `NaN` values are treated as equal. This does not incur a high cost on the implementation side where the equality and hash operation would simply gain a special case for `NaN`. 

The use case we are seeing is that sometimes we get floating point inputs to our models which are essentially categorical encodings (including `NaN`s). We use label encoders to map those floating point categories to integers or strings but don't want to necessarily map `NaN`s to the "default/unseen" case.

Reply:
Is this a resolved conversation?